antarest/study/business/link/CompositeLinkDAO.py

```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2024, RTE (https://www.rte-france.com)
+#
+# See AUTHORS.txt
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# SPDX-License-Identifier: MPL-2.0
+#
+# This file is part of the Antares project.
+
+from typing import List
+
+from antarest.study.business.link.LinkDAO import LinkDAO
+from antarest.study.business.model.link_model import LinkDTO
+from antarest.study.model import Study
+
+
+class CompositeLinkDAO(LinkDAO):
+    """
+    CompositeLinkDAO acts as a composite data access object (DAO) for managing study links.
+    It delegates operations to two underlying DAOs: one for cache and one for persistent storage.
+    This class ensures that the cache and storage are kept in sync during operations.
+
+    Attributes:
+        cache_dao (LinkDAO): The DAO responsible for managing links in the cache.
+        storage_dao (LinkDAO): The DAO responsible for managing links in persistent storage.
+    """
+
+    def __init__(self, cache_dao: LinkDAO, storage_dao: LinkDAO):
+        """
+        Initializes the CompositeLinkDAO with a cache DAO and a storage DAO.
+
+        Args:
+            cache_dao (LinkDAO): DAO for managing links in the cache.
+            storage_dao (LinkDAO): DAO for managing links in persistent storage.
+        """
+        self.cache_dao = cache_dao
+        self.storage_dao = storage_dao
+
+    def get_all_links(self, study: Study) -> List[LinkDTO]:
+        """
+        Retrieves all links for a given study.
+
+        This method first tries to retrieve links from the cache. If the cache is empty,
+        it fetches the links from the persistent storage, updates the cache with the retrieved links,
+        and then returns the list of links.
+
+        Args:
+            study (Study): The study for which to retrieve the links.
+
+        Returns:
+            List[LinkDTO]: A list of all links associated with the study.
+        """
+        links = self.cache_dao.get_all_links(study)
+        if not links:
```

Comment:
Or maybe a third alternative !

- the cache implementation could throw a `CacheMissError`, that we will catch here 

It will rely on a quite "hidden" convention but why not ! Relying on exceptions is quite pythonic from what I know ...